### PR TITLE
RA-1148 - Search concept handler is now able to filter search by class

### PR DESCRIPTION
**Pull request by tmarzeion**

### CODE REVIEW

File: omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs1_8/ConceptSearchHandler1_8.java [hunk 1]
```diff
@@ -47,7 +48,18 @@
 	ConceptService conceptService;
 	
 	private final SearchConfig searchConfig = new SearchConfig("default", RestConstants.VERSION_1 + "/concept", Arrays.asList("1.8.*", "1.9.*", "1.10.*", "1.11.*", "1.12.*", "2.0.*"),
-	        Arrays.asList(new SearchQuery.Builder("Allows you to find concepts by source and code").withRequiredParameters("source").withOptionalParameters("code").build(),new SearchQuery.Builder("Allows you to find concepts by name").withRequiredParameters("name").build()));
```
Comment (tmarzeion) [inline omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs1_8/ConceptSearchHandler1_8.java]: Small code cleanup, I think it looks pro now

**Comment by dkayiwa** [comment]
@tmarzeion can you add some unit tests for this?

**Comment by tmarzeion** [comment]
@dkayiwa Tests created

**Comment by tmarzeion** [comment]
@dkayiwa How is code review going? :)

File: omod-common/src/main/java/org/openmrs/module/webservices/rest/web/resource/api/SearchQuery.java [hunk 2]
```diff
@@ -43,12 +43,20 @@ public Builder(String description) {
 		}
 		
 		public Builder withRequiredParameters(String... requiredParameters) {
+			if (searchQuery.requiredParameters != null) {
+				throw new IllegalStateException("requiredParameters() must not be called twice");
```
Comment (dkayiwa) [inline omod-common/src/main/java/org/openmrs/module/webservices/rest/web/resource/api/SearchQuery.java]: Do you have a test for this too?

File: omod-common/src/main/java/org/openmrs/module/webservices/rest/web/resource/api/SearchQuery.java [hunk 3]
```diff
@@ -43,12 +43,20 @@ public Builder(String description) {
...
+			}
+			
 			searchQuery.requiredParameters = Collections.unmodifiableSet(new HashSet<String>(Arrays
 			        .asList(requiredParameters)));
 			return this;
 		}
 		
 		public Builder withOptionalParameters(String... optionalParameters) {
+			if (searchQuery.optionalParameters != null) {
+				throw new IllegalStateException("withOptionalParameters() must not be called twice");
```
Comment (dkayiwa) [inline omod-common/src/main/java/org/openmrs/module/webservices/rest/web/resource/api/SearchQuery.java]: Do you have a test for this too?

File: omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs1_8/ConceptSearchHandler1_8.java [hunk 4]
```diff
@@ -65,32 +79,56 @@ public PageableResult search(RequestContext context) throws ResponseException {
...
-				
-				return new NeedsPaging<Concept>(concepts, context);
-			}
-			else
-			{
-				return new EmptySearchResult();
 			}
 		}
+		else {
+			throw new InvalidSearchException("Invalid searchType: " + searchType + ". Allowed values: \"equals\" and \"fuzzy\"");
```
Comment (dkayiwa) [inline omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs1_8/ConceptSearchHandler1_8.java]: Do you have a test for this too?

File: omod-1.9/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs1_9/ConceptController1_8Test.java [hunk 5]
```diff
@@ -212,6 +212,27 @@ public void shouldReturnTheAuditInfoForTheFullRepresentation() throws Exception
...
+
+	@Test
+	public void shouldSearchAndReturnConceptsFilteredByClassAndName() throws Exception {
+		MockHttpServletRequest req = request(RequestMethod.GET, getURI());
+		SimpleObject result;
+		List<Object> hits;
+
+		String conceptClassUuid = "3d065ed4-b0b9-4710-9a17-6d8c4fd259b7"; // DRUG
+		String name = "Aspirin"; //ASPIRIN
+		String searchType = "equals";
```
Comment (dkayiwa) [inline omod-1.9/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs1_9/ConceptController1_8Test.java]: Does this test alone cater for the two if and else paths that you have introduced above?

File: omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs1_8/ConceptSearchHandler1_8.java (hunk 4 above)
Comment (tmarzeion) [inline omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs1_8/ConceptSearchHandler1_8.java]: Done

File: omod-common/src/main/java/org/openmrs/module/webservices/rest/web/resource/api/SearchQuery.java (hunk 3 above)
Comment (tmarzeion) [inline omod-common/src/main/java/org/openmrs/module/webservices/rest/web/resource/api/SearchQuery.java]: Done

File: omod-common/src/main/java/org/openmrs/module/webservices/rest/web/resource/api/SearchQuery.java (hunk 2 above)
Comment (tmarzeion) [inline omod-common/src/main/java/org/openmrs/module/webservices/rest/web/resource/api/SearchQuery.java]: Done at shouldThrowExceptionWhenSearchRequiredParametersAreCalledTwice()

File: omod-1.9/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs1_9/ConceptController1_8Test.java (hunk 5 above)
Comment (tmarzeion) [inline omod-1.9/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs1_9/ConceptController1_8Test.java]: @dkayiwa I've improved test coverage, check it out :)

**Comment by rkorytkowski** [comment]
@dkayiwa, it looks good to me now. Would you like to have another look and merge?

**Pull request merged**